Template README refactor (#267)
fredrikhgrelland/vagrant-hashistack: template/README.md

<p align="center">
 <img width="100px" src="https://www.svgrepo.com/show/58111/cube.svg" align="center" alt="Vagrant-hashistack" />
 <h2 align="center">Vagrant-hashistack Template</h2>
 <p align="center">Starter template for `fredrikhgrelland/vagrant-hashistack`</p>
<p align="center">
    <a href="https://github.com/fredrikhgrelland/vagrant-hashistack-template/actions">
      <img alt="Build" src="https://github.com/fredrikhgrelland/vagrant-hashistack-template/workflows/CI/CD/badge.svg" />
    </a>

>     <a href="https://github.com/fredrikhgrelland/vagrant-hashistack-template/releases">

Author: That's a good point. But not all releases of vagrant-hashistack constitute a release of template. I will add this change for now

         1. [Consul Secrets Engine](#consul-secrets-engine)
4. [Usage](#usage)
   1. [Commands](#commands)
   2. [MinIO](#minio)
      1. [Pushing Resources To MinIO With Ansible (Docker image)](#pushing-resources-to-minio-with-ansible-docker-image)
      2. [Fetching Resources From MinIO With Nomad (Docker image)](#fetching-resources-from-minio-with-nomad-docker-image)
   3. [Iteration of the Development Process](#iteration-of-the-development-process)
5. [Test Configuration and Execution](#test-configuration-and-execution)

## Description - What & Why
This template is a starting point, and example, on how to take advantage of the [Hashistack vagrant-box](https://app.vagrantup.com/fredrikhgrelland/boxes/hashistack) to create, develop, and test Terraform-modules within the Hashistack ecosystem.

**Hashistack**, in current repository context, is a set of software products by [HashiCorp](https://www.hashicorp.com/).

On github, you may use the
<a href="https://github.com/fredrikhgrelland/vagrant-hashistack-template/generate" alt="Use this template">
<img src="https://img.shields.io/badge/Use this template-2ea44f"/></a>
 button to generate a new repository from this template.

> > :bulb: If you found this in `fredrikhgrelland/vagrant-hashistack`, you may be interested in the separate repository [vagrant-hashistack-template](https://github.com/fredrikhgrelland/vagrant-hashistack-template/).

Author: What do you mean? You might be reading that sentence from `vagrant-hashistack` and not the `vagrant-hashistack-template` repo

### Why Does This Exist?

>  This template aims to nudge you in the direction of a well thought out workflow for building terraform-modules, with a full framework for testing, using the `fredrikhgrelland/hashistack` vagrant-box.

Author: I like this one!

## Install Prerequisites

```text
make install
```

The command, will install:
- [VirtualBox](https://www.virtualbox.org/)
- [Packer](https://www.packer.io/)
- [Vagrant](https://www.vagrantup.com/) with additional plugins
- [Additional software dependent on the OS (Linux, MacOS)](../install/Makefile)

### Packages that needs to be pre-installed

- [Make](https://man7.org/linux/man-pages/man1/make.1.html)
- [Git CLI](https://git-scm.com/book/en/v2/Getting-Started-The-Command-Line)

#### MacOS Specific
- Virtualization must be enabled. [This is enabled by default on MacOS.](https://support.apple.com/en-us/HT203296)
- [Homebrew](https://brew.sh/) must be installed.

#### Ubuntu Specific
- Virtualization must be enabled. [Error if it is not.](https://github.com/fredrikhgrelland/vagrant-hashistack/issues/136)
- Packages [gpg](http://manpages.ubuntu.com/manpages/xenial/man1/gpg.1.html) and [apt](http://manpages.ubuntu.com/manpages/bionic/man8/apt.8.html) must be installed.

---

`NB` _Post installation you might need to reboot your system in order to start the virtual-provider (VirtualBox)_

---

## Configuration

> ### Default Startup Scheme

Author: I agree, fixed

> From a thousand foot view the default startup scheme will:
> 1. Start the hashistack and MinIO
> 2. Run [playbook.yml](dev/ansible/playbook.yml), which in turn runs all ansible-playbooks inside [dev/ansible/](dev/ansible).
> 
> > :bulb: Vagrantfile lines 8-11 run the first playbook on startup, and can be changed.

Reviewer: It is a bit controversial to me, see the value of this `thousand foot view` description. I would suggest jumping onto `detailed` one here.
```suggestion
```

Author: I like it because it is a very rough description, missing lots of nuances, but maybe a little too "casual" of an explanation. Any other opinions @fredrikhgrelland @Neha-Sinha2305?

> :bulb: Below is a detailed description of the _whole_ startup procedure, both user changeable and not.

---

#### Detailed Startup Procedure
_Preloaded_ - Comes bundled with the box, not possible to change

_System provided_ - Provided by the system in automated processes, not possible to change

_User provided_ - Provided by the user to alter the box or template in some way

1. `/home/vagrant/.env_default` - _preloaded_ - default variables
2. `vagrant/.env` - _user provided_ - variables override, see [Pre-packaged Configuration Switches](#pre-packaged-configuration-switches) for details
3. `vagrant/.env_override` - _system provided_ - variables are overridden for test purposes

> 4. `vagrant/dev/vagrant/conf/pre_ansible.sh` - _user provided_ - script running before ansible bootstrap procedure, [details](dev/vagrant/conf/pre_bootstrap/README.md)

Author: It's actually correct that it is not present, because this is a user provided file. There are details about this in the README

> 4. `vagrant/dev/vagrant/conf/pre_ansible.sh` - _user provided_ - script running before ansible bootstrap procedure, [details](dev/vagrant/conf/pre_bootstrap/README.md)

Author: Fixed in latest commit

5. `vagrant/dev/vagrant/conf/pre_bootstrap/*.yml` - _user provided_ - pre bootstrap tasks, running before hashistack software starts, [details](dev/vagrant/conf/README.md)
6. `/etc/ansible/bootstrap.yml` - _preloaded_ - verify ansible variables and software configuration, run hashistack software and MinIO, & verify that it started correctly
7. `vagrant/conf/post_bootstrap/*.yml` - _user provided_ - poststart scripts, running after hashistack software has started, [details](dev/vagrant/conf/pre_bootstrap/README.md)

> 8. `vagrant/dev/conf/pre_ansible.sh` - _user provided_ - script running after ansible bootstrap procedure, [details](dev/vagrant/conf/README.md)

Author: What is wrong?

> 8. `vagrant/dev/conf/pre_ansible.sh` - _user provided_ - script running after ansible bootstrap procedure, [details](dev/vagrant/conf/README.md)

Reviewer 2: It should be ```/vagrant/dev/vagrant/conf/post_ansible.sh```

9. `vagrant/ansible/*.yml` - _user provided_ - ansible tasks included in playbook, see [Pre-packaged Configuration Switches](#pre-packaged-configuration-switches) for details

---

> 

Reviewer: I would suggest better formatting here and a diagram. 

Table formating try

```suggestion
||||
|-------------|:-------------|:-----|
|1 `/home/vagrant/.env_default`|[ _preloaded_ ]| default variables |
|2 `vagrant/.env`|[ _user provided_ ]| variables override, see [Pre-packaged Configuration Switches](#pre-packaged-configuration-switches) for details |
|3 `vagrant/.env_override`|[ _system provided_ ]| variables are overridden for test purposes |
|4 `vagrant/dev/vagrant/conf/pre_ansible.sh`|[ _user provided_ ]| script running before ansible bootstrap procedure, [details](dev/vagrant/conf/pre_bootstrap/README.md) |
|5 `vagrant/dev/vagrant/conf/pre_bootstrap/*.yml`|[ _user provided_ ]| pre bootstrap tasks, running before hashistack software starts, [details](dev/vagrant/conf/README.md) |
|6 `/etc/ansible/bootstrap.yml`|[ _preloaded_ ]| verify ansible variables and software configuration, run hashistack software and MinIO, & verify that it started correctly |
|7 `vagrant/conf/post_bootstrap/*.yml`|[ _user provided_ ]| poststart scripts, running after hashistack software has started, [details](dev/vagrant/conf/pre_bootstrap/README.md) |
|8 `vagrant/dev/conf/pre_ansible.sh`|[ _user provided_ ]| script running after ansible bootstrap procedure, [details](dev/vagrant/conf/README.md) |
|9 `vagrant/ansible/*.yml`|[ _user provided_ ]| ansible tasks included in playbook, see [Pre-packaged Configuration Switches](#pre-packaged-configuration-switches) for details |
```

Author: Fixed

### Pre and Post Hashistack Startup Procedure
You may change the hashistack configuration or add additional pre and post steps to the startup procedure to match your needs.
Detailed documentation in [dev/vagrant/conf/README.md](dev/vagrant/conf/README.md)

> ### Pre-packaged Configuration Switches

Reviewer: Just `NB`
We probably need to follow one name for such functionality. In root documentation, the name has not been explicitly chosen (toggles, switches, ...). If we go for switches, it is better to name them the same everywhere.
Link in root https://github.com/fredrikhgrelland/vagrant-hashistack#configuration

> The box comes standard with a set of environment switches to simplify testing of different scenarios and enable staged development efforts.

Reviewer: ```suggestion
The box comes [with a set of environment switches](https://github.com/fredrikhgrelland/vagrant-hashistack#configuration) to simplify testing of different scenarios and enable staged development efforts.
```

To change any of these values from their defaults, you may add the environment variable to [.env](dev/.env).

> NB: All lowercase variables will automatically get a corresponding  `TF_VAR_` prepended variant for use directly in terraform.

Reviewer: Can you provide a link to the script for prepending functionality  (`TF_VAR_` prepended variant) ?

Author: Done

> :bulb: Below is a detailed description of the _whole_ startup procedure, both user changeable and not.

---

#### Detailed Startup Procedure
_Preloaded_ - Comes bundled with the box, not possible to change

_System provided_ - Provided by the system in automated processes, not possible to change

_User provided_ - Provided by the user to alter the box or template in some way

1. `/home/vagrant/.env_default` - _preloaded_ - default variables
2. `/vagrant/.env` - _user provided_ - variables override, see [Pre-packaged Configuration Switches](#pre-packaged-configuration-switches) for details
3. `/vagrant/.env_override` - _system provided_ - variables are overridden for test purposes
4. `/vagrant/dev/vagrant/conf/pre_ansible.sh` - _user provided_ - script running before ansible bootstrap procedure, [details](dev/vagrant/conf/README.md)
5. `/vagrant/dev/vagrant/conf/pre_bootstrap/*.yml` - _user provided_ - pre bootstrap tasks, running before hashistack software starts, [details](dev/vagrant/conf/README.md)
6. `/etc/ansible/bootstrap.yml` - _preloaded_ - verify ansible variables and software configuration, run hashistack software and MinIO, & verify that it started correctly

> 7. `/vagrant/conf/post_bootstrap/*.yml` - _user provided_ - poststart scripts, running after hashistack software has started, [details](dev/vagrant/conf/pre_bootstrap/README.md)

Reviewer 2: ```suggestion
7. `/vagrant/dev/vagrant/conf/post_bootstrap/*.yml` - _user provided_ - poststart scripts, running after hashistack software has started, [details](dev/vagrant/conf/post_bootstrap/README.md)
```

<!-- markdownlint-disable MD041 -->
<p align="center">
 <img width="100px" src="https://www.svgrepo.com/show/58111/cube.svg" align="center" alt="Vagrant-hashistack" />
 <h2 align="center">Vagrant-hashistack Template</h2>
 <p align="center">Starter template for <a href="https://github.com/fredrikhgrelland/vagrant-hashistack">fredrikhgrelland/vagrant-hashistack</a></p>
<p align="center">
    <a href="https://github.com/fredrikhgrelland/vagrant-hashistack-template/actions">
      <img alt="Build" src="https://github.com/fredrikhgrelland/vagrant-hashistack-template/workflows/CI/CD/badge.svg" />
    </a>
    <a href="https://github.com/fredrikhgrelland/vagrant-hashistack/releases">

>       <img alt="Releases" src="https://img.shields.io/github/v/release/fredrikhgrelland/vagrant-hashistack-template?label=latest%20version" />

Reviewer: Wrong path?
<img width="656" alt="Screenshot 2020-08-25 at 16 12 15" src="https://user-images.githubusercontent.com/15572799/91185205-cf6ebc80-e6ed-11ea-8c3f-644c5e6ae342.png">

Author: Not the wrong path, but I'll be honest I'm not quite sure how the release process works for the template-repo. If they are directly coupled with vagrant-hashistack releases, or if they will be their own separate thing. Would be nice if there was a tag for template-releases that creates a separate release in the template repo

Reviewer: Ref #270

Author: I'm resolving this for now, followup in #270

>       <img alt="Releases" src="https://img.shields.io/github/v/release/fredrikhgrelland/vagrant-hashistack-template?label=latest%20version" />

Reviewer 2: Should we do this temporary fix so as we fetch the release version by pointing to ```vagrant-hashistack```, that would look better than the text ```no release or repo not found```?
And of course we can change this when we would have solved #270?
```suggestion
      <img alt="Releases" src="https://img.shields.io/github/v/release/fredrikhgrelland/vagrant-hashistack?label=latest%20version" />
```

Author: Yeah, sounds good, let's do that

    </a>
    <a href="https://github.com/fredrikhgrelland/vagrant-hashistack/commits">
      <img alt="Updated" src="https://img.shields.io/github/last-commit/fredrikhgrelland/vagrant-hashistack-template?label=last%20updated" />
    </a>
    <br />
    <br />
    <p align="center">
      <a href="https://github.com/fredrikhgrelland/vagrant-hashistack-template/generate" alt="Clone Template">
            <img src="https://img.shields.io/badge/Github-Clone%20template-blue?style=for-the-badge&logo=github" />
        </a>
    </p>
</p>

> 

Reviewer: Just an offer (if will be applied, do NOT forget to remove `/` escape char)
```suggestion
## Start

To start to develop terraform module:

/```text
make dev
/```

Follow [Iteration of the development process](#iteration-of-the-development-process) for more information.
```

Author: Hm, I agree with this, but there will be a full "how-to" later, so I think it's better to put it in that PR

Reviewer: Could you create an issue and resolve this conversation? @pdmthorsrud

Author: Yup, will do in like 15 minutes, my day is a little packed!

Author: Added #272, resolving

- Packages [gpg](http://manpages.ubuntu.com/manpages/xenial/man1/gpg.1.html) and [apt](http://manpages.ubuntu.com/manpages/bionic/man8/apt.8.html) must be installed.

---

`NB` _Post installation you might need to reboot your system in order to start the virtual-provider (VirtualBox)_

---

## Configuration

### Startup Scheme
From a thousand foot view the startup scheme will:
1. Start the hashistack and MinIO
2. Run [playbook.yml](dev/ansible/playbook.yml), which in turn runs all ansible-playbooks inside [dev/ansible/](dev/ansible).

> :bulb: Vagrantfile lines 8-11 run the first playbook on startup, and can be changed.

> :bulb: Below is a detailed description of the _whole_ startup procedure, both user changeable and not.

---

#### Detailed Startup Procedure
_box_ - Comes bundled with the box, not possible to change

_system_ - Provided by the system in automated processes, not possible to change

_user_ - Provided by the user to alter the box or template in some way

| What | Provided by | Description |
|:------------|:------------:|:-----|
|1 `/home/vagrant/.env_default`|[ _box_ ]| default variables |

> |2 `vagrant/.env`|[ _user_ ]| variables override, see [Pre-packaged Configuration Switches](file:///home/m88614/projects/vagrant-hashistack/template/README.md#pre-packaged-configuration-switches) for details |

Reviewer: typo in link

Author: fixed!

> |3 `vagrant/.env_override`|[ _system_ ]| variables are overridden for test purposes |

Reviewer: Why some of the paths are absolute, some relative? (Just noticed it now. )

Author: Huh.. I thought I changed this. I _swear_ I changed it to `/vagrant`.

Might be that I copy-pasted the initial table-version. Anyways, I'll make them absolute again!

Reviewer: Mine example was not correct :)

Author: Fixed!

|4 `/vagrant/dev/vagrant/conf/pre_ansible.sh`|[ _user_ ]| script running before ansible bootstrap procedure, [details](dev/vagrant/conf/pre_bootstrap/README.md) |
|5 `/vagrant/dev/vagrant/conf/pre_bootstrap/*.yml`|[ _user_ ]| pre bootstrap tasks, running before hashistack software starts, [details](dev/vagrant/conf/README.md) |
|6 `/etc/ansible/bootstrap.yml`|[ _box_ ]| verify ansible variables and software configuration, run hashistack software and MinIO, & verify that it started correctly,  [link](../ansible/bootstrap.yml) |
|7 `/vagrant/conf/post_bootstrap/*.yml`|[ _user_ ]| poststart scripts, running after hashistack software has started, [details](dev/vagrant/conf/pre_bootstrap/README.md) |

> |8 `/vagrant/dev/conf/pre_ansible.sh`|[ _user_ ]| script running after ansible bootstrap procedure, [details](dev/vagrant/conf/README.md) |

Reviewer: ```suggestion
|8 `/vagrant/dev/conf/post_ansible.sh`|[ _user_ ]| script running after ansible bootstrap procedure, [details](dev/vagrant/conf/README.md) |
```

Author: Dang, good catch!

> | What | Provided by | Description |
> |:------------|:------------:|:-----|
> |1 `/home/vagrant/.env_default`|[ _box_ ]| default variables |
> |2 `/vagrant/.env`|[ _user_ ]| variables override, see [Pre-packaged Configuration Switches](#pre-packaged-configuration-switches) for details |
> |3 `/vagrant/.env_override`|[ _system_ ]| variables are overridden for test purposes |
> |4 `/vagrant/dev/vagrant/conf/pre_ansible.sh`|[ _user_ ]| script running before ansible bootstrap procedure, [details](dev/vagrant/conf/pre_bootstrap/README.md) |
> |5 `/vagrant/dev/vagrant/conf/pre_bootstrap/*.yml`|[ _user_ ]| pre bootstrap tasks, running before hashistack software starts, [details](dev/vagrant/conf/README.md) |
> |6 `/etc/ansible/bootstrap.yml`|[ _box_ ]| verify ansible variables and software configuration, run hashistack software and MinIO, & verify that it started correctly,  [link](../ansible/bootstrap.yml) |
> |7 `/vagrant/conf/post_bootstrap/*.yml`|[ _user_ ]| poststart scripts, running after hashistack software has started, [details](dev/vagrant/conf/pre_bootstrap/README.md) |
> |8 `/vagrant/dev/conf/post_ansible.sh`|[ _user_ ]| script running after ansible bootstrap procedure, [details](dev/vagrant/conf/README.md) |
> |9 `/vagrant/ansible/*.yml`|[ _user_ ]| ansible tasks included in playbook, see [Pre-packaged Configuration Switches](#pre-packaged-configuration-switches) for details |

Reviewer: ```suggestion
|Seq number| What | Provided by | Description |
|:--:|:------------|:------------:|:-----|
|1 |`/home/vagrant/.env_default`|[ _box_ ]| default variables |
|2 |`/vagrant/.env`|[ _user_ ]| variables override, see [Pre-packaged Configuration Switches](#pre-packaged-configuration-switches) for details |
|3 |`/vagrant/.env_override`|[ _system_ ]| variables are overridden for test purposes |
|4 |`/vagrant/dev/vagrant/conf/pre_ansible.sh`|[ _user_ ]| script running before ansible bootstrap procedure, [details](dev/vagrant/conf/pre_bootstrap/README.md) |
|5 |`/vagrant/dev/vagrant/conf/pre_bootstrap/*.yml`|[ _user_ ]| pre bootstrap tasks, running before hashistack software starts, [details](dev/vagrant/conf/README.md) |
|6 |`/etc/ansible/bootstrap.yml`|[ _box_ ]| verify ansible variables and software configuration, run hashistack software and MinIO, & verify that it started correctly,  [link](../ansible/bootstrap.yml) |
|7 |`/vagrant/conf/post_bootstrap/*.yml`|[ _user_ ]| poststart scripts, running after hashistack software has started, [details](dev/vagrant/conf/pre_bootstrap/README.md) |
|8 |`/vagrant/dev/conf/post_ansible.sh`|[ _user_ ]| script running after ansible bootstrap procedure, [details](dev/vagrant/conf/README.md) |
|9 |`/vagrant/ansible/*.yml`|[ _user_ ]| ansible tasks included in playbook, see [Pre-packaged Configuration Switches](#pre-packaged-configuration-switches) for details |
```

Author: Ah yeah, that's a good idea!

> |9 `vagrant/ansible/*.yml`|[ _user_ ]| ansible tasks included in playbook, see [Pre-packaged Configuration Switches](file:///home/m88614/projects/vagrant-hashistack/template/README.md#pre-packaged-configuration-switches) for details |

Reviewer: typo in link

Author: fixed!

---

### Pre and Post Hashistack Startup Procedure
#### Ansible Playbooks Pre and Post Hashistack Startup
You may change the hashistack configuration or add additional pre and post steps to the ansible startup procedure to match your needs.
Detailed documentation in [dev/vagrant/conf/README.md](dev/vagrant/conf/README.md)

#### Bash Scripts Pre and Post Ansible Playbook

> In addition to ansible playbooks you can also add bash-scripts that will be run before and after the ansible provisioning step. This is useful for doing deeper changes to the box pertaining your needs. Detailed documentation in [dev/vagrant/conf/README.md](dev/vagrant/conf/README.md)

Reviewer: ```suggestion
In addition to ansible playbooks, you can also add bash-scripts that will be run before and/or after the ansible provisioning step. This is useful for doing deeper changes to the box pertaining to your needs. Detailed documentation in [dev/vagrant/conf/README.md](dev/vagrant/conf/README.md)
```

### Pre-packaged Configuration Switches

> The box comes [with a set of environment switches](https://github.com/fredrikhgrelland/vagrant-hashistack#configuration) to simplify testing of different scenarios and enable staged development efforts.

Reviewer: ```suggestion
The box comes [with a set of configuration switches controlled by env variables](https://github.com/fredrikhgrelland/vagrant-hashistack#configuration) to simplify testing of different scenarios and enable staged development efforts.
```

Reviewer: Refer to #269

vagrant ssh -c 'vault read consul/creds/admin-team'
```

> > :bulb: Tokens can be used to access UI (different access level depends on role)

Reviewer: ```suggestion
> :bulb: Tokens can be used to access UI (different access level depends on policy attached to the token)
```

## Usage
### Commands
There are several commands that help to run the vagrant-box:
- `make install` installs all prerequisites

- `make up` provisions a [vagrant-hashistack](https://github.com/fredrikhgrelland/vagrant-hashistack/) box on your machine. After the machine and hashistack are set up it will run the [Default Startup Scheme](#default-startup-scheme).

- `make clean` takes down the provisioned box if there is any.

- `make update` downloads the newest version of the [vagrant-hashistack box](https://github.com/fredrikhgrelland/vagrant-hashistack/) from [vagrantcloud](https://vagrantcloud.com/fredrikhgrelland/hashistack).

- `make example` runs the example in [template_example/](template_example)

> :bulb: For full info, check `template/Makefile`.

> > :warning: Makefile commands are not idempotent in the context of vagrant-box.  You could face the error of port collisions. Most of the cases it could happen because of the vagrant box has already been running.

Reviewer 2: We need to specify the remedy as well to get around this problem/situation...like run ```make clean``` or run 
```
vagrant global-status
vagrant destroy -f <id> 
```

Author: Fixed in latest commit, thanks!

vagrant ssh -c 'vault read consul/creds/admin-team'
```

> :bulb: Tokens can be used to access UI (different access level depends on role)

## Usage
### Commands
There are several commands that help to run the vagrant-box:

> - `make install` installs all prerequisites

Reviewer: I would suggest notifying the user, that he/she does not need to run it every time.
```suggestion
- `make install` installs all prerequisites. Run once.
```

> - `make up` provisions a [vagrant-hashistack](https://github.com/fredrikhgrelland/vagrant-hashistack/) box on your machine. After the machine and hashistack are set up it will run the [Default Startup Scheme](#default-startup-scheme).

Reviewer: ```suggestion
- `make up` provisions a [vagrant-hashistack](https://github.com/fredrikhgrelland/vagrant-hashistack/) box on your machine, deploy Terraform module, and run all tests locally.
```

Author: Don't agree with this one. It does not necessarily deploy the terraform module, that's up to the user (at least the way we have it now).

Reviewer: I guess ref to #243

- `make clean` takes down the provisioned box if there is any.

- `make update` downloads the newest version of the [vagrant-hashistack box](https://github.com/fredrikhgrelland/vagrant-hashistack/) from [vagrantcloud](https://vagrantcloud.com/fredrikhgrelland/hashistack).

- `make example` runs the example in [template_example/](template_example)

> 

Reviewer: ```suggestion
`make dev` - the same as `make up`, but skip all tests.
```

Author: This is not true. Skips only https://github.com/fredrikhgrelland/vagrant-hashistack/blob/dcab1d24280e28bf2639a2b16545e73069ce83f3/ansible/bootstrap.yml#L11-L15 atm

Reviewer: What the differences then or point to have `up` and `dev`?

Author: Right now there is no point (which is why I created issue #243).

Reviewer: Ok. we will fix it in another PR

> > :bulb: For full info, check `template/Makefile`.

Reviewer: ```suggestion
> :bulb: For full info, check [`template/Makefile`](./Makefile).
```

vagrant ssh -c 'vault read consul/creds/admin-team'
```

> :bulb: Tokens can be used to access UI (different access level depends on policy attached to the token)

## Usage
### Commands
There are several commands that help to run the vagrant-box:
- `make install` installs all prerequisites

- `make up` provisions a [vagrant-hashistack](https://github.com/fredrikhgrelland/vagrant-hashistack/) box on your machine. After the machine and hashistack are set up it will run the [Startup Scheme](#startup-scheme).

- `make clean` takes down the provisioned box if there is any.

- `make update` downloads the newest version of the [vagrant-hashistack box](https://github.com/fredrikhgrelland/vagrant-hashistack/) from [vagrantcloud](https://vagrantcloud.com/fredrikhgrelland/hashistack).

- `make example` runs the example in [template_example/](template_example)

> :bulb: For full info, check `template/Makefile`.
> :warning: Makefile commands are not idempotent in the context of vagrant-box.  You could face the error of port collisions. Most of the cases it could happen because of the vagrant box has already been running. Run `vagrant destroy -f` to destroy the box.

> Once vagrant-box is running, you can use other options like the Nomad- and Terraform-CLIs to iterate over the deployment in the development stage.

Reviewer: ```suggestion
Once vagrant-box is running, you can use other [options like the Nomad- and Terraform-CLIs to iterate over the deployment in the development stage](#iteration-of-the-development-process).
```

nomad job run countdash.hcl
```

> :warning: _Your local binaries and the binaries in the box might not be the same versions, and may behave differently. [Box versions.](../ansible/group_vars/all/variables.yml)

2. **Using vagrant**. Box instance has all binaries are installed and available in the PATH.

> You can use ´vagrant ssh´ to place yourself inside of the vagrantbox and run commands.

Reviewer: ```suggestion
You can use `vagrant ssh` to place yourself inside of the vagrantbox and run commands.
```
